### PR TITLE
use typescript, add signed header request using postmessage

### PR DESCRIPTION
**Pull request by lenkan**
This PR updates the client implementation significantly.

- Does not subscribe to messages when importing, but rather after a "createClient" call.
- Adds type information to all requests.
- Adding tests,
- Adding eslint
- Adding formatting.

Changing the API to 3 calls.

- "authorize" - for the web app to send an "authorize" request to the extension. See docs in implementation.
- "signData" - sends a request to sign data to the extension. It is up to the extension to whether or not it should require manual approval.
- "signRequest" - sends a request to sign a HTTP request to the extension. The extension should sign the request information and return the signify headers.

### CODE REVIEW

**Comment by 2byrds** [comment]
@rodolfomiranda and @HunnySajid please take a look when you get a chance

File: src/client.ts
Comment (HunnySajid) [inline src/client.ts:196]: Currently in polaris-web, the api allows authorization with either credential (`requestCredential`: shows list of credentials to select one), aid (`requestAid`: shows list of aid to select one), or both (`requestAidORCred`: shows list of both creds and aids). We might need to add another method to distinguish this behavior (e.g `authorizeWithAid`).

Regarding `requestAutoSignin`:
There exists a feature named autoSign, the idea is to make a user select a sign in pair (combination or webapp origin + (aid OR credential)) with autoSignin flag as `true` (One origin can only have one signin pair with autoSignin true). So, the moment users request `requestAutoSignin`, they are not shown a list of credential/aid instead that particular signin pair gets selected. We'd also need to add this functionality.

Comment (rodolfomiranda) [reply src/client.ts:196]: Is the `signData` intended to sign any arbitrary data? We should try to avoid requesting signature of any data for security concerns. It may be better to have specific method when requesting signatures so the extension can alert the user and request consent why such a signature is requested. Sign a request is one case of a specific signature. Others can be added, just as sign a document.

Comment (lenkan) [reply src/client.ts:196]: Thanks for responding!

@rodo Yes. Currently the functionality is to sign arbitrary data. It is up to the extension implementation whether or not to ask user for consent before the signing is executed though. In our implementation the promise returned from signData from is resolved or rejected then based on user consent.

We also thought of having the user provide the data/documents in the extension, but the user journey was tricky to solve I think. It's up for grabs though... I agree with you rodo that we should avoid exposing a method to sign arbitrary data.

@hunni, thank you for clarifying! As you can see with a lot of the recent commits, we broke the compatibility. For example the autosignin. The plan is to finalise a minimum API surface that we are happy with, then try re-add compatibility again. Hopefully we are able to align.

In addition, we were thinking about requiring explicit user approval before injecting the content script. This would also make Google happier and probably speed up the review process because you don't need host permissions.

File: src/client.ts
Comment (HunnySajid) [inline src/client.ts:30]: @lenkan I would appreciate it if you could explain how is this implemented this sessionId in your extension. 
Is it made by combining an identifier for the tab e.g sender.tab.id? 
Or it works if web app is opened in multiple tabs? It would be good for us to implement this in signify-browser-extension.

**Review by HunnySajid — APPROVED** [review]
With just a few missing/breaking changes (i'll add soon in upcoming PR's), it is a significant improvement. LGTM!

File: src/client.ts
Comment (lenkan) [reply src/client.ts:30]: Sorry for the delay @HunnySajid. I have just gotten back to this now and I think the sessionId string should actually be removed from the API here. The purpose of it was to be able to handle multiple tabs/sessions at the same time but isolated from each other. 

However, I think the solution is a bit flawed. I would rather do what you suggested with `sender.tab.id` to keep track of the sender. In that implementation, the session id does not have to be part of the polaris API. It would just be an implementation detail of the browser extension. I am going to try and make those changes ASAP.

Comment (HunnySajid) [reply src/client.ts:30]: Thanks @lenkan 
this way the session is associated with a distinguish tab instead of a webapp (or domain if you will). No need to share sessionId with webapp.

Comment (lenkan) [reply src/client.ts]: @HunnySajid I have updated the API now to remove the "sessionId" parameter.

I believe the things we need to land this is to re-introduce compatibility with the signify browser extension. However, for that, I think we either need to do some sort of "handshake" to determine if "externally_connectable" is enabled, or change the implementation in signify-browser-extension to based on post message for all features.

Comment (HunnySajid) [reply src/client.ts]: > @HunnySajid I have updated the API now to remove the "sessionId" parameter.
> 
> I believe the things we need to land this is to re-introduce compatibility with the signify browser extension. However, for that, I think we either need to do some sort of "handshake" to determine if "externally_connectable" is enabled, or change the implementation in signify-browser-extension to based on post message for all features.

@lenkan thank you for making the relevant changes. I have been lately busy with compatible changes on signify-browser-extension side (was working with [source branch](https://github.com/nordlei/polaris-web/tree/signed-headers) of this pr). The externally_connetable is omitted now, postmessage is used instead. As soon as this PR is merged i'll create a couple of subsequent PRs on browser extension side for compatibility.

**Review by rodolfomiranda — APPROVED** [review]
LGTM

**Pull request merged**
